ext/opentelemetry-ext-pyramid/src/opentelemetry/ext/pyramid/callbacks.py

```diff
@@ -0,0 +1,170 @@
+    _insert_tween(config)
+
+
+def _insert_tween(config):
+    settings = config.get_settings()
+    tweens = settings.get("pyramid.tweens")
+    # If the list is empty, pyramid does not consider the tweens have been
+    # set explicitly. And if our tween is already there, nothing to do
+    if not tweens or not tweens.strip():
+        # Add our tween just before the default exception handler
+        config.add_tween(TWEEN_NAME, over=EXCVIEW)
+
+
+def _before_traversal(event):
+    request = event.request
+    environ = request.environ
+    span_name = otel_wsgi.get_default_span_name(environ)
+
+    enabled = environ.get(_ENVIRON_ENABLED_KEY)
+    if enabled is None:
+        _logger.warning(
+            "Opentelemetry pyramid tween 'opentelemetry.ext.pyramid.trace_tween_factory'"
+            "was not called. Make sure that the tween is included in 'pyramid.tweens' if"
+            "the tween list was created manually"
+        )
+        return
+
+    if not enabled:
+        # Tracing not enabled, return
+        return
+
+    start_time = environ.get(_ENVIRON_STARTTIME_KEY)
+
+    token = context.attach(
+        propagators.extract(otel_wsgi.get_header_from_environ, environ)
+    )
+    tracer = trace.get_tracer(__name__, __version__)
+
+    attributes = otel_wsgi.collect_request_attributes(environ)
+
+    if request.matched_route:
+        span_name = request.matched_route.pattern
+        attributes["http.route"] = request.matched_route.pattern
+    else:
+        span_name = otel_wsgi.get_default_span_name(environ)
+
+    span = tracer.start_span(
+        span_name,
+        kind=trace.SpanKind.SERVER,
+        attributes=attributes,
+        start_time=start_time,
+    )
+
+    activation = tracer.use_span(span, end_on_exit=True)
+    activation.__enter__()
+    environ[_ENVIRON_ACTIVATION_KEY] = activation
+    environ[_ENVIRON_SPAN_KEY] = span
+    environ[_ENVIRON_TOKEN] = token
+
+
+def trace_tween_factory(handler, registry):
+    settings = registry.settings
+    enabled = asbool(settings.get(SETTING_TRACE_ENABLED, True))
+
+    if enabled:
+        # make a request tracing function
+        def trace_tween(request):
+            if disable_trace(request.url, _excluded_hosts, _excluded_paths):
+                request.environ[_ENVIRON_ENABLED_KEY] = False
+                # short-circuit when we don't want to trace anything
+                return handler(request)
+
+            request.environ[_ENVIRON_ENABLED_KEY] = True
+            request.environ[_ENVIRON_STARTTIME_KEY] = time_ns()
+
+            response = None
+            try:
+                response = handler(request)
+            except HTTPException as exc:
+                # If the exception is a pyramid HTTPException,
+                # that's still valuable information that isn't necessarily
+                # a 500. For instance, HTTPFound is a 302.
+                # As described in docs, Pyramid exceptions are all valid
+                # response types
```

Comment:
I think in both cases the exception would be raised - we raise at the end of the except, which calls the finally block to record to the span and then the exception is reraised

Reply:
I probably don't understand try..except..finally with raise + return at all but I ran a small test and it seems the exception is not re-raised if the finally block contains a return statement. 

Also, if the intention is for the exception to be re-raised, I'd suggest to assign the exception to something other than the response to make it the intention clear.

Reply:
https://repl.it/@OwaisLone1/ElectricImperturbableListener#main.py test but may be I'm doing something wrong?

Reply:
right, if you have a return in the finally it clobbers the exception. However we don't here, the return is outside the finally block (I can definitely see how its hard to tell, I should add a new line before the return). I can see what you mean by not assigning the exception to the response, but it lets us re-use the code in the finally block since the HTTPException has all of the properties necessary to fill out the span. Up for suggestions if you can think of a better way.

Reply:
How about assigning both response from the try block and the exception to a 3rd var name `result` or `response_or_exception` (or something else) and using that to fill out the span? That would make it very clear enough I think.

Reply:
changed to response_or_exception, let me know if it looks better now (I think it does)

Reply:
Yes, more obvious for sure. I actually meant we catch response in the `response` variable but also assign it to `response_or_exception`. That way response var will be returned and `response_or_exception` will be used to extract metadata. Just a suggestion :)

Reply:
That makes a lot more sense :P updated to return response and get span attributes from response_or_exception